src/Grpc.AspNetCore.Server/Internal/PercentEncodingHelpers.cs

```diff
@@ -0,0 +1,218 @@
+            {
+                return value;
+            }
+
+            // Encode
+            return string.Create(encodedLength, value, Encode);
+
+            static void Encode(Span<char> span, string s)
+            {
+                Span<byte> unicodeBytesBuffer = stackalloc byte[MaxUnicodeCharsReallocate * MaxUtf8BytesPerUnicodeChar];
+
+                var writePosition = 0;
+                for (var i = 0; i < s.Length; i++)
+                {
+                    var current = s[i];
+                    if (current > AsciiMaxValue)
+                    {
+                        // Get additional unicode characters
+                        var unicodeCharCount = GetUnicodeCharCount(s, i);
+
+                        var numberOfBytes = Encoding.UTF8.GetBytes(s.AsSpan(i, unicodeCharCount), unicodeBytesBuffer);
+
+                        for (var count = 0; count < numberOfBytes; count++)
+                        {
+                            EscapeAsciiChar(ref span, ref writePosition, (char)unicodeBytesBuffer[count]);
+                        }
+                        i += unicodeCharCount - 1;
+                    }
+                    else if (IsUnreservedCharacter(current))
+                    {
+                        span[writePosition++] = current;
+                    }
+                    else
+                    {
+                        EscapeAsciiChar(ref span, ref writePosition, current);
+                    }
+                }
+            }
+        }
+
+        private static void EscapeAsciiChar(ref Span<char> span, ref int writePosition, char current)
+        {
+            span[writePosition++] = '%';
+            span[writePosition++] = HexChars[current >> 4];
+            span[writePosition++] = HexChars[current & 15];
+        }
+
+        private static int GetUnicodeCharCount(string value, int currentIndex)
+        {
+            var unicodeCharCount = 0;
+            var maxSize = Math.Min(value.Length - currentIndex, MaxUnicodeCharsReallocate - 1); // Leave a character for possible low surrogate
+            for (; unicodeCharCount < maxSize && value[currentIndex + unicodeCharCount] > AsciiMaxValue; unicodeCharCount++)
+            {
+            }
+
+            if (char.IsHighSurrogate(value[currentIndex + unicodeCharCount - 1]))
+            {
+                if (unicodeCharCount < value.Length - currentIndex && char.IsLowSurrogate(value[currentIndex + unicodeCharCount]))
+                {
+                    // Last character is a high surrogate so check ahead to see if it is followed by a low surrogate and include
+                    unicodeCharCount++;
+                }
+            }
+
+            return unicodeCharCount;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool IsUnreservedCharacter(char c)
+        {
+            return ((PercentEncodingUnreservedBitField[c / 8] >> (c % 8)) & 1) != 0;
```

Comment:
is doing all this arithmetic really better than just a 128 long bool array?

Reply:
If it changes, it will change to something like discussed at https://github.com/grpc/grpc-dotnet/pull/223#discussion_r279640292